rollout/sync.go

```diff
@@ -857,6 +857,38 @@ func (c *rolloutContext) resetRolloutStatus(newStatus *v1alpha1.RolloutStatus) {
 	newStatus.CurrentStepIndex = replicasetutil.ResetCurrentStepIndex(c.rollout)
 }
 
+func (c *rolloutContext) isRollbackWithinWindow() bool {
+	if c.newRS == nil || c.stableRS == nil {
+		return false
+	}
+	// first check if this is a rollback
+	if c.newRS.CreationTimestamp.Before(&c.stableRS.CreationTimestamp) {
```

Comment:
In playing around with this there is interesting behavior when going back and forward within a couple of replica sets because if we are rolling back to a replicaset that exists already it does not update the created timestamp. I wonder if it makes more sense to use this annotation `rollout.argoproj.io/revision: "9"`  to determine if we are within range instead. It dose change the behavior slightly in that it would not let you jump back multiple "revisions" which also could be a good thing.

Reply:
We had a similar question internally, but I would prefer to stick with the creation timestamp, because it normally follows the build order. Jumping back too many builds is not an issue, as:
- this is only meant to be used during "incidents", when it's critical to get to a previously known good known version as fast as possible
- we are anyway limited by the revisionHistoryLimit which is a number between 5-10

If we used the revision annotation instead of the creation timestamp, then the build order is not taken into account any more, and I can imagine a scenario where we fast rollback to a build that is "newer" than the current stable one (because the revision is mutated every time a build is promoted).

For instance, imagine:
- build 1 - rev 1
- build 2 - rev 2
- build 3 - rev 3
- build 4 - rev 4
window size: 2

with Revision Annotation:
- rollback to 2 (fast), now build 2 - rev 5 (history becomes: 1, 3, 4, 5)
- rollback to 3 (fast), now build 3 - rev 6 (history becomes: 1, 4, 5, 6) - this is already wrong, as build 3 is "newer" than build 2, so it may contain changes that broke the deployment

with Creation timestamp:
- rollback to 2 (fast)
- rollback to 3 (regular deployment)

Reply:
Yup that makes sense and thank you for the detailed explanation.